### PR TITLE
Use go:embed for presets and reconcile VirtualMachine instancetype and preference

### LINKED ISSUE #11024: Deploy kubermatic-standard instancetype and preference

### PR DESCRIPTION
**Pull request by hdurand0710**
Signed-off-by: Helene Durand <helene@kubermatic.com>

**What this PR does / why we need it**:
This PR goals:
- use go:embed and yaml manifests for KubeVirt static resources instead of the currently hard-coded VirtualMachineInstancePreset, and place the manifests under `/resources/` folder.
- use it for the currently deployed `kubermatic-standard` VirtualMachineInstancePreset
- use it to deploy 2 new resources:
  - `kubermatic-standard`  VirtualMachineInstancetype (that will replace the kubermatic-standard` VirtualMachineInstancePreset)
  - `kubermatic-standard`  VirtualMachinePreference

The mechanism will reconcile into the infrastructure KubeVirt cluster all manifests:
- pkg/provider/cloud/kubevirt/manifests/instancetypes
- pkg/provider/cloud/kubevirt/manifests/preferences
- pkg/provider/cloud/kubevirt/manifests/presets

**Which issue(s) this PR fixes**:

Fixes #11024
This PR is the first step for 10645.

**What type of PR is this?**

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
Introduction of KubeVirt default instance types and instance preferences that will replace virtual machine presets
```

**Documentation**:

```documentation
TBD. Documentation will be done later when last PR for the Epic 10645 is raised.
```

### CODE REVIEW

**Review by mfranczy — CHANGES_REQUESTED** [review]
I think we should not add kubevirt manifests to `resources/cloudcontroller` pkg. The pkg should contain  the CCM logic only. `InstanceType` etc. is KubeVirt infra specific. Currently I don't see any better place than `provider/cloud/kubevirt`.

File: hack/update-codegen.sh [hunk 1]
```diff
@@ -19,7 +19,7 @@ set -euo pipefail
 cd $(dirname $0)/..
 source hack/lib.sh
 
-CONTAINERIZE_IMAGE=golang:1.19.1 containerize ./hack/update-codegen.sh
+CONTAINERIZE_IMAGE=quay.io/kubermatic/build:go-1.19-node-18-1 containerize ./hack/update-codegen.sh
```
Comment (mfranczy) [inline hack/update-codegen.sh:22]: @xrstf could you confirm that this the image we want to use?

File: pkg/provider/cloud/kubevirt/manifests/manifests.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+func (s *StandardPreferenceGetter) GetManifestFS() *ManifestFS {
+	return &ManifestFS{
+		Fs:     standardPreferenceFs,
+		Folder: standardPreferenceFolder,
+	}
+}
+
+// KubernetesFromYaml returns a list of Kubernetes runtime objects from their yaml templates.
+// It returns the objects for all files included in the ManifestFS folder, skipping (with error log) the yaml files
+// that would not correct yaml files.
```
Comment (mfranczy) [inline pkg/provider/cloud/kubevirt/manifests/manifests.go]: > that would not contain correct yaml files. 

:)

File: pkg/provider/cloud/kubevirt/manifests/manifests.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+func KubernetesFromYaml(client ctrlruntimeclient.Client, manifestFsGetter ManifestFSGetter) []runtime.Object {
```
Comment (mfranczy) [inline pkg/provider/cloud/kubevirt/manifests/manifests.go]: let's maybe rename to
`RuntimeFromYaml`?

File: pkg/provider/cloud/kubevirt/manifests/manifests.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	//go:embed preset
+	standardPresetsFs    embed.FS
+	standardPresetFolder = "preset"
```
Comment (mfranczy) [inline pkg/provider/cloud/kubevirt/manifests/manifests.go]: let's keep all names plural:
* presets
* instancetypes
* preferences

File: pkg/provider/cloud/kubevirt/manifests/manifests.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	standardInstancetypeFolder = "instancetype"
+
+	//go:embed preference
+	standardPreferenceFs     embed.FS
+	standardPreferenceFolder = "preference"
+)
+
+type ManifestFS struct {
+	Fs     embed.FS
+	Folder string
```
Comment (mfranczy) [inline pkg/provider/cloud/kubevirt/manifests/manifests.go]: Let's change `Folder` to `Dir`

File: pkg/provider/cloud/kubevirt/manifests/manifests_test.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+var (
+	fakeclient ctrlruntimeclient.Client
+	//go:embed test/presetOk
+	testPresetFS     embed.FS
+	testPresetFolder = path.Join("test", "presetOk")
+
+	//go:embed test/presetOkKo
+	tesOkKoPresetFS      embed.FS
+	testOkKoPresetFolder = path.Join("test", "presetOkKo")
```
Comment (mfranczy) [inline pkg/provider/cloud/kubevirt/manifests/manifests_test.go]: as discussed let's remove `presetOkKo` and keep only one directory for all preset fixtures. The invalid one you can prefix with `invalid-` and later check in the code. Additionaly please add invalid resource for preferences and instancetypes as well.

**Review by mfranczy — CHANGES_REQUESTED** [review]
Looks good, just a few minor things

File: pkg/provider/cloud/kubevirt/manifests/instancetype/kubermatic-standard.yaml [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: instancetype.kubevirt.io/v1alpha1
+kind: VirtualMachineInstancetype
+metadata:
+  name: kubermatic-standard
```
Comment (mfranczy) [inline pkg/provider/cloud/kubevirt/manifests/instancetype/kubermatic-standard.yaml]: Except presets, let's remove kubermatic prefix from each instancetypes and preferences

**Comment by mfranczy** [comment]
@hdurand0710 for InstanceTypes let's go with
* standard-2 - 2 vCPU 8Gi memory
* standard-4 - 4 vCPU 16Gi memory
* standard-8 - 8 vCPU 32Gi memory

and for preferences
* sockets-advantage - where we map all cpu units to sockets.

**Comment by mfranczy** [comment]
Additionaly don't forget about resource quota :), maybe that should be another PR? We have to keep backward compatibiliy of calculating presets.

File: hack/update-codegen.sh (hunk 1 above)
Comment (xrstf) [reply hack/update-codegen.sh:22]: Yup. 👍

**Comment by hdurand0710** [comment]
> Additionaly don't forget about resource quota calculation :), maybe that should be another PR? We have to keep backward compatibiliy of calculating presets.

yop, that will be in the API PR, the next one.

**Comment by hdurand0710** [comment]
> @hdurand0710 for InstanceTypes let's go with
> 
> * standard-2 - 2 vCPU 8Gi memory
> * standard-4 - 4 vCPU 16Gi memory
> * standard-8 - 8 vCPU 32Gi memory
> 
> and for preferences
> 
> * sockets-advantage - where we map all cpu units to sockets.
> 
> For the API endpoint category let's change `standard` to `kubermatic`. Thanks!

I will do it in the next PR dealing with API endpoints. Thanks!

File: pkg/provider/cloud/kubevirt/manifests/manifests.go (hunk 4 above)
Comment (hdurand0710) [reply pkg/provider/cloud/kubevirt/manifests/manifests.go]: done

File: pkg/provider/cloud/kubevirt/manifests/manifests.go (hunk 5 above)
Comment (hdurand0710) [reply pkg/provider/cloud/kubevirt/manifests/manifests.go]: done

File: pkg/provider/cloud/kubevirt/manifests/manifests_test.go (hunk 6 above)
Comment (hdurand0710) [reply pkg/provider/cloud/kubevirt/manifests/manifests_test.go]: done

File: pkg/provider/cloud/kubevirt/manifests/instancetype/kubermatic-standard.yaml (hunk 7 above)
Comment (hdurand0710) [reply pkg/provider/cloud/kubevirt/manifests/instancetype/kubermatic-standard.yaml]: done

**Review by mfranczy — APPROVED** [review]
/approve
/lgtm

**Comment by mfranczy** [comment]
@hdurand0710 could you as well make the release note one liner? I think we don't have to go so deep to inform where resources exist, that information should be part of documentation.

I think it's enough to say something like this:
> Introduction of KubeVirt default instance types and instance preferences that will replace virtual machine presets

**Review by ahmedwaleedmalik — APPROVED** [review]
/approve

**Pull request merged**